### PR TITLE
Make Malinka Async

### PR DESCRIPTION
**Pull request by jojojames**
See 
https://github.com/LefterisJP/malinka/issues/12

Should fix things like this where shell-command-to-string hangs for some reason (as well as blocks the UI for a few seconds even in the good case).

```
;; This was retrieved by setting toggle-debug-on-quit to t and C-g'ing while Emacs was stalled.

Debugger entered--Lisp error: (quit)
  call-process("/bin/zsh" nil t nil "-c" "/Users/james/.emacs.d/elpa/rtags-20170408.1642/rtags-2.9/bin/rc -w")
  apply(call-process "/bin/zsh" nil t nil ("-c" "/Users/james/.emacs.d/elpa/rtags-20170408.1642/rtags-2.9/bin/rc -w"))
  process-file("/bin/zsh" nil t nil "-c" "/Users/james/.emacs.d/elpa/rtags-20170408.1642/rtags-2.9/bin/rc -w")
  shell-command-to-string("/Users/james/.emacs.d/elpa/rtags-20170408.1642/rtags-2.9/bin/rc -w")
  (progn (shell-command-to-string cmd))
  (if rc (progn (shell-command-to-string cmd)))
  (let* ((rc (rtags-executable-find "rc")) (cmd (s-join " " (cons rc args)))) (if rc (progn (shell-command-to-string cmd))))
  (progn (let* ((rc (rtags-executable-find "rc")) (cmd (s-join " " (cons rc args)))) (if rc (progn (shell-command-to-string cmd)))))
  (if (malinka--rtags-assert-rdm-runs) (progn (let* ((rc (rtags-executable-find "rc")) (cmd (s-join " " (cons rc args)))) (if rc (progn (shell-command-to-string cmd))))))
  malinka--rtags-invoke-with("-w")
  (let* ((output (malinka--rtags-invoke-with "-w")) (project-root (progn "Access slot \"root-directory\" of `malinka--project' struct CL-X." nil (or (progn nil (and (memq ... cl-struct-malinka--project-tags) t)) (signal (quote wrong-type-argument) (list (quote malinka--project) project))) (aref project 2))) (loadedlist (s-match (format "%s.*" project-root) output))) loadedlist)
  malinka--rtags-project-known\?(#s(malinka--project "testcode2" "~/Code/testcodeworld/c/testcode/" "~/Code/testcodeworld/c/testcode/build/" nil "cmake .. -DCMAKE_BUILD_TYPE=Debug -DHEADLESS=1" "if [ ! -d ~/Code/testcodeworld/c/testcode/build/ ]; then mkdir -p ~/Code/testcodeworld/c/testcode/build/ && cd ~/Code/testcodeworld/c/testcode/build/ && cmake .. -DCMAKE_BUILD_TYPE=Debug -DHEADLESS=1; else cd ~/Code/testcodeworld/c/testcode/build/; fi; make -j4" nil nil nil))
  (not (malinka--rtags-project-known\? project))
  (cond ((and (eq fileattr (quote not-configured)) (not (malinka--project-compatible-cmake\? project)) (not (malinka--rtags-is-indexing\?)) (not (malinka--project-being-configured\? project))) (if malinka-print-info\? (progn (message (concat "Malinka-info: " "Project \"%s\" does not seem to be configured. Configuring ...") (malinka--project-name project)))) (malinka--project-map-update-compiledb project)) ((not (malinka--rtags-project-known\? project)) (if malinka-print-info\? (progn (message (concat "Malinka-info: " "Rtags does not know about \"%s\". Informing it.") (malinka--project-name project)))) (malinka--project-create-or-select-compiledb project)) ((not (malinka--rtags-project-loaded\? project)) (if malinka-print-info\? (progn (message (concat "Malinka-info: " "Rtags knows about \"%s\" but does not have it loaded. Loading it.") (malinka--project-name project)))) (malinka--try-select-project project)))
  (let ((project (nth 0 query)) (fileattr (nth 1 query))) (malinka--rtags-assert-rdm-runs) (cond ((and (eq fileattr (quote not-configured)) (not (malinka--project-compatible-cmake\? project)) (not (malinka--rtags-is-indexing\?)) (not (malinka--project-being-configured\? project))) (if malinka-print-info\? (progn (message (concat "Malinka-info: " "Project \"%s\" does not seem to be configured. Configuring ...") (malinka--project-name project)))) (malinka--project-map-update-compiledb project)) ((not (malinka--rtags-project-known\? project)) (if malinka-print-info\? (progn (message (concat "Malinka-info: " "Rtags does not know about \"%s\". Informing it.") (malinka--project-name project)))) (malinka--project-create-or-select-compiledb project)) ((not (malinka--rtags-project-loaded\? project)) (if malinka-print-info\? (progn (message (concat "Malinka-info: " "Rtags knows about \"%s\" but does not have it loaded. Loading it.") (malinka--project-name project)))) (malinka--try-select-project project))))
  (progn (let ((project (nth 0 query)) (fileattr (nth 1 query))) (malinka--rtags-assert-rdm-runs) (cond ((and (eq fileattr (quote not-configured)) (not (malinka--project-compatible-cmake\? project)) (not (malinka--rtags-is-indexing\?)) (not (malinka--project-being-configured\? project))) (if malinka-print-info\? (progn (message (concat "Malinka-info: " "Project \"%s\" does not seem to be configured. Configuring ...") (malinka--project-name project)))) (malinka--project-map-update-compiledb project)) ((not (malinka--rtags-project-known\? project)) (if malinka-print-info\? (progn (message (concat "Malinka-info: " "Rtags does not know about \"%s\". Informing it.") (malinka--project-name project)))) (malinka--project-create-or-select-compiledb project)) ((not (malinka--rtags-project-loaded\? project)) (if malinka-print-info\? (progn (message (concat "Malinka-info: " "Rtags knows about \"%s\" but does not have it loaded. Loading it.") (malinka--project-name project)))) (malinka--try-select-project project)))))
  (if query (progn (let ((project (nth 0 query)) (fileattr (nth 1 query))) (malinka--rtags-assert-rdm-runs) (cond ((and (eq fileattr (quote not-configured)) (not (malinka--project-compatible-cmake\? project)) (not (malinka--rtags-is-indexing\?)) (not (malinka--project-being-configured\? project))) (if malinka-print-info\? (progn (message ... ...))) (malinka--project-map-update-compiledb project)) ((not (malinka--rtags-project-known\? project)) (if malinka-print-info\? (progn (message ... ...))) (malinka--project-create-or-select-compiledb project)) ((not (malinka--rtags-project-loaded\? project)) (if malinka-print-info\? (progn (message ... ...))) (malinka--try-select-project project))))))
  (let* ((filename (buffer-file-name buffer)) (query (malinka--file-belongs-to-project filename))) (if query (progn (let ((project (nth 0 query)) (fileattr (nth 1 query))) (malinka--rtags-assert-rdm-runs) (cond ((and (eq fileattr ...) (not ...) (not ...) (not ...)) (if malinka-print-info\? (progn ...)) (malinka--project-map-update-compiledb project)) ((not (malinka--rtags-project-known\? project)) (if malinka-print-info\? (progn ...)) (malinka--project-create-or-select-compiledb project)) ((not (malinka--rtags-project-loaded\? project)) (if malinka-print-info\? (progn ...)) (malinka--try-select-project project)))))))
  (progn (let* ((filename (buffer-file-name buffer)) (query (malinka--file-belongs-to-project filename))) (if query (progn (let ((project (nth 0 query)) (fileattr (nth 1 query))) (malinka--rtags-assert-rdm-runs) (cond ((and ... ... ... ...) (if malinka-print-info\? ...) (malinka--project-map-update-compiledb project)) ((not ...) (if malinka-print-info\? ...) (malinka--project-create-or-select-compiledb project)) ((not ...) (if malinka-print-info\? ...) (malinka--try-select-project project))))))))
  (if (malinka--buffer-is-c\? buffer) (progn (let* ((filename (buffer-file-name buffer)) (query (malinka--file-belongs-to-project filename))) (if query (progn (let ((project ...) (fileattr ...)) (malinka--rtags-assert-rdm-runs) (cond (... ... ...) (... ... ...) (... ... ...))))))))
  (let ((buffer (current-buffer))) (if (malinka--buffer-is-c\? buffer) (progn (let* ((filename (buffer-file-name buffer)) (query (malinka--file-belongs-to-project filename))) (if query (progn (let (... ...) (malinka--rtags-assert-rdm-runs) (cond ... ... ...))))))))
  malinka--idle-project-check()
  apply(malinka--idle-project-check nil)
  timer-event-handler([t 0 1 0 t malinka--idle-project-check nil idle 0])
```

### CODE REVIEW

**Comment by jojojames** [comment]
I'm not sure how to run your tests by the way but it's working smoothly as far as I'm concerned.

**Comment by LefterisJP** [comment]
@jojojames Thank you very much for this. I initially just made malinka as a way for me to easily work with rtags, but as people seem to be using it this is a step in the good direction.

RE tests: They are too old, so old in fact I don't remember anything about them. Should make proper tests and make a travis build. Will make an issue for this.

Until then I will test this branch manually in my machine and if it works fine and smoothly I will merge.

**Comment by jojojames** [comment]
Sounds good, I'm putting a lot of use into Malinka right now too so I'm dogfooding my pull also.

Let me know if you see any problems, I'll be glad to fix them.

**Comment by LefterisJP** [comment]
> I'm putting a lot of use into Malinka right now too 

This makes me really happy :)

One general comment: PRs should be self-contained and deal with only one thing.
I see you did a lot of style changes even though the project has no defined style guidelines.

I don't dislike the usage of `elisp-formatter`, I just have never used it before and would prefer that being another PR if we decide to go with that.

File: malinka.el [hunk 1]
```diff
@@ -1,4 +1,4 @@
-;;; malinka.el --- A C/C++ project configuration package for Emacs
+;;; malinka.el --- A C/C++ project configuration package for Emacs -*- lexical-binding: t; -*-
```
Comment (LefterisJP) [inline malinka.el]: When I try to `byte-compile` the file, having `lexical-binding` as `t` destroys everything and it does not seem to work.
Why did you add lexical binding to the file?

File: malinka.el [hunk 2]
```diff
@@ -843,28 +854,72 @@ Returns the output of the command as a string or nil in case of error"
...
+     (let* ((loadedlist (s-match "\\(.*\\) <=" output))
+            (dir (when loadedlist (nth 1 loadedlist))))
+       (when dir
+         (funcall
+          callback (f-equal? dir (malinka--project-root-directory project))))))
+   "-w"))
+
+(defun async-shell-command-to-string (command callback)
+  "Execute shell command COMMAND asynchronously in the
+  background.
```
Comment (LefterisJP) [inline malinka.el]: why is the `background` in its own line here and why unlike the following lines it has 2 spaces before it?

Comment (jojojames) [inline malinka.el]: Mistake, though I'm not sure why the formatter doesn't format the string.

**Comment by jojojames** [comment]
Split the pulls up and closing this one.

File: malinka.el (hunk 1 above)
Comment (jojojames) [inline malinka.el]: I made a change to that in the other pull request. (Using push instead of add-to-list)

Dynamic binding was making it a little tricky to debug the callbacks (the names kept colliding together).

https://www.reddit.com/r/emacs/comments/5j9qgz/lexicalbinding_t_vs_lexicallet/
